### ISSUE TITLE
`prefer_void_to_null` should not be reported in a Cast Pattern or an As Expression

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#4201]
`void` is not allowed in either:

```dart
void f(int a) {
  switch (a) {
    case var _ as void: // compile-time error
  }
  a as void; // compile-time error
}
```

So `prefer_void_to_null` should not be reported in those positions.
